File: tests/chess-engine.test.js
'use strict';
const {
    CH_WHITE, CH_BLACK, CH_PAWN, CH_KNIGHT, CH_BISHOP, CH_ROOK, CH_QUEEN, CH_KING,
    newChessGame, getLegalMoves, applyMove, updateStatus,
    isInCheck, findKing, evaluateBoard, getChessAIMove, rawMoves
} = require('../js/chess-engine');

describe('newChessGame', () => {
    test('returns correct structure', () => {
        const g = newChessGame();
        expect(g).toHaveProperty('board');
        expect(g).toHaveProperty('turn');
        expect(g).toHaveProperty('castling');
        expect(g).toHaveProperty('enPassant');
        expect(g).toHaveProperty('status');
    });
    test('board is 8×8', () => {
        const { board } = newChessGame();
        expect(board.length).toBe(8);
        board.forEach(row => expect(row.length).toBe(8));
    });
    test('white pawns on row 6', () => {
        const { board } = newChessGame();
        board[6].forEach(p => expect(p).toBe(CH_PAWN * CH_WHITE));
    });
    test('black pawns on row 1', () => {
        const { board } = newChessGame();
        board[1].forEach(p => expect(p).toBe(CH_PAWN * CH_BLACK));
    });
    test('white king at row 7 col 4', () => {
        const { board } = newChessGame();
        expect(board[7][4]).toBe(CH_KING * CH_WHITE);
    });
    test('black king at row 0 col 4', () => {
        const { board } = newChessGame();
        expect(board[0][4]).toBe(CH_KING * CH_BLACK);
    });
    test('initial turn is white', () => {
        expect(newChessGame().turn).toBe(CH_WHITE);
    });
    test('all castling rights start true', () => {
        const { castling } = newChessGame();
        expect(castling).toEqual({ wK: true, wQ: true, bK: true, bQ: true });
    });
    test('no en passant initially', () => {
        expect(newChessGame().enPassant).toBeNull();
    });
});

describe('getLegalMoves', () => {
    test('white has 20 legal moves at game start', () => {
        const g = updateStatus(newChessGame());
        expect(g.legalMoves.length).toBe(20);
    });
    test('all moves have from and to fields', () => {
        const g = updateStatus(newChessGame());
        g.legalMoves.forEach(m => {
            expect(typeof m.r).toBe('number');
            expect(typeof m.c).toBe('number');
            expect(typeof m.tr).toBe('number');
            expect(typeof m.tc).toBe('number');
        });
    });
    test('no moves in checkmate position', () => {
        // Fool's mate: e4, f5, Qh5, g5, Qxf7# or Scholar's mate position
        let g = updateStatus(newChessGame());
        // Manually set a checkmate board
        const mateBoard = Array.from({ length: 8 }, () => new Array(8).fill(0));
        mateBoard[7][4] = CH_KING * CH_WHITE;  // white king e1
        mateBoard[0][4] = CH_KING * CH_BLACK;  // black king e8
        mateBoard[1][3] = CH_QUEEN * CH_WHITE; // white queen d7 — mates black
        mateBoard[1][5] = CH_QUEEN * CH_WHITE; // white queen f7 — double attack
        const mateState = updateStatus({ board: mateBoard, turn: CH_BLACK,
            castling:{wK:false,wQ:false,bK:false,bQ:false}, enPassant:null,
            halfMove:0, fullMove:1, status:'active' });
        // Black might not be in checkmate but should have very few moves
        expect(mateState.legalMoves.length).toBeLessThanOrEqual(5);
    });
});

describe('applyMove', () => {
    test('pawn moves forward', () => {
        const g  = updateStatus(newChessGame());
        const m  = g.legalMoves.find(mv => mv.r===6 && mv.c===4 && mv.tr===5);
        const g2 = applyMove(g, m);
        expect(g2.board[5][4]).toBe(CH_PAWN * CH_WHITE);
        expect(g2.board[6][4]).toBe(0);
    });
    test('double pawn push sets en passant', () => {
        const g  = updateStatus(newChessGame());
        const m  = g.legalMoves.find(mv => mv.double);
        const g2 = applyMove(g, m);
        expect(g2.enPassant).not.toBeNull();
    });
    test('turn switches after move', () => {
        const g  = updateStatus(newChessGame());
        const m  = g.legalMoves[0];
        expect(applyMove(g, m).turn).toBe(CH_BLACK);
    });
    test('does not mutate original state', () => {
        const g   = updateStatus(newChessGame());
        const orig = JSON.stringify(g.board);
        applyMove(g, g.legalMoves[0]);
        expect(JSON.stringify(g.board)).toBe(orig);
    });
    test('rook loses castling rights when moved', () => {
        const g = { ...newChessGame(),
            board: newChessGame().board.map(r => r.slice()),
            castling: { wK:true, wQ:true, bK:true, bQ:true } };
        // Clear space to move rook
        g.board[7][1] = 0; g.board[7][2] = 0; g.board[7][3] = 0;
        const state = updateStatus(g);
        const rookMove = state.legalMoves.find(m => m.r===7 && m.c===0);
        if (!rookMove) return;
        const g2 = applyMove(state, rookMove);
        expect(g2.castling.wQ).toBe(false);
    });
});

describe('isInCheck', () => {
    test('not in check at game start', () => {
        const g = newChessGame();
        expect(isInCheck(g.board, CH_WHITE)).toBe(false);
        expect(isInCheck(g.board, CH_BLACK)).toBe(false);
    });
    test('detects check from queen', () => {
        const board = Array.from({ length: 8 }, () => new Array(8).fill(0));
        board[7][4] = CH_KING   * CH_WHITE;
        board[4][4] = CH_QUEEN  * CH_BLACK;   // attacks along file
        board[0][4] = CH_KING   * CH_BLACK;
        expect(isInCheck(board, CH_WHITE)).toBe(true);
    });
    test('blocked check is not check', () => {
        const board = Array.from({ length: 8 }, () => new Array(8).fill(0));
        board[7][4] = CH_KING  * CH_WHITE;
        board[5][4] = CH_PAWN  * CH_WHITE;   // blocker
        board[4][4] = CH_QUEEN * CH_BLACK;
        board[0][4] = CH_KING  * CH_BLACK;
        expect(isInCheck(board, CH_WHITE)).toBe(false);
    });
});

describe('findKing', () => {
    test('finds white king at initial position', () => {
        const k = findKing(newChessGame().board, CH_WHITE);
        expect(k).toEqual({ r: 7, c: 4 });
    });
    test('finds black king at initial position', () => {
        const k = findKing(newChessGame().board, CH_BLACK);
        expect(k).toEqual({ r: 0, c: 4 });
    });
    test('returns null when king missing', () => {
        const board = Array.from({ length: 8 }, () => new Array(8).fill(0));
        expect(findKing(board, CH_WHITE)).toBeNull();
    });
});

describe('evaluateBoard', () => {
    test('initial position evaluates near 0', () => {
        const score = evaluateBoard(newChessGame().board);
        expect(Math.abs(score)).toBeLessThan(100);  // should be very close to 0
    });
    test('board with extra white queen favours white', () => {
        const board = newChessGame().board.map(r => r.slice());
        board[4][4] = CH_QUEEN * CH_WHITE;
        expect(evaluateBoard(board)).toBeGreaterThan(0);
    });
    test('board with extra black queen favours black', () => {
        const board = newChessGame().board.map(r => r.slice());
        board[4][4] = CH_QUEEN * CH_BLACK;
        expect(evaluateBoard(board)).toBeLessThan(0);
    });
});

describe('getChessAIMove', () => {
    test('returns a move at game start for each depth (one per unique depth)', () => {
        const g = updateStatus(newChessGame());
        // CHESS_DEPTH_MAP = [0,1,1,2,2,2,3,3,4,4] — test one representative per unique depth
        for (const d of [0, 1, 3, 6, 8]) {
            expect(getChessAIMove(g, d)).not.toBeNull();
        }
    }, 10000);

    test('returns null when no legal moves', () => {
        const board = Array.from({ length: 8 }, () => new Array(8).fill(0));
        board[0][4] = CH_KING * CH_BLACK;
        board[0][3] = CH_QUEEN * CH_WHITE; board[0][5] = CH_QUEEN * CH_WHITE;
        board[1][4] = CH_ROOK  * CH_WHITE; board[7][4] = CH_KING  * CH_WHITE;
        const state = updateStatus({ board, turn: CH_BLACK,
            castling:{wK:false,wQ:false,bK:false,bQ:false},
            enPassant:null, halfMove:0, fullMove:1, status:'active' });
        if (state.status === 'checkmate' || state.status === 'stalemate') {
            expect(getChessAIMove(state, 1)).toBeNull();
        }
    });
});

describe('updateStatus', () => {
    test('sets status to active at game start', () => {
        const g = updateStatus(newChessGame());
        expect(g.status).toBe('active');
    });
    test('attaches legalMoves array', () => {
        const g = updateStatus(newChessGame());
        expect(Array.isArray(g.legalMoves)).toBe(true);
    });
});

File: kotlin/shared/src/commonTest/kotlin/xyz/zyxwonderland/shared/chess/ChessEngineTest.kt
package xyz.zyxwonderland.shared.chess

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class ChessEngineTest {

    // ── Initial state ──────────────────────────────────────────────────────────

    @Test fun `newChessGame starts white to move`() {
        assertEquals(CH_WHITE, newChessGame().turn)
    }

    @Test fun `newChessGame board has correct piece counts`() {
        val board = newChessGame().board
        var whitePawns = 0; var blackPawns = 0
        for (r in 0..7) for (c in 0..7) {
            if (board[r][c] == CH_PAWN)  whitePawns++
            if (board[r][c] == -CH_PAWN) blackPawns++
        }
        assertEquals(8, whitePawns)
        assertEquals(8, blackPawns)
    }

    @Test fun `newChessGame all castling rights set`() {
        val c = newChessGame().castling
        assertTrue(c.wK); assertTrue(c.wQ); assertTrue(c.bK); assertTrue(c.bQ)
    }

    @Test fun `newChessGame has no en passant`() {
        assertNull(newChessGame().enPassant)
    }

    @Test fun `newChessGame status is ACTIVE`() {
        assertEquals(ChessStatus.ACTIVE, newChessGame().status)
    }

    // ── Legal moves ────────────────────────────────────────────────────────────

    @Test fun `white has 20 legal moves at start`() {
        assertEquals(20, getLegalMoves(newChessGame()).size)
    }

    @Test fun `pawn can advance two squares from starting row`() {
        val moves = getLegalMoves(newChessGame())
        val e2e4  = moves.any { it.r == 6 && it.c == 4 && it.tr == 4 && it.tc == 4 }
        assertTrue(e2e4)
    }

    @Test fun `knight moves are correct from starting position`() {
        val moves    = getLegalMoves(newChessGame())
        val knightMoves = moves.filter { it.r == 7 && it.c == 1 } // Nb1
        assertEquals(2, knightMoves.size)
    }

    // ── applyMove ─────────────────────────────────────────────────────────────

    @Test fun `applyMove e2-e4 sets en passant`() {
        val state = newChessGame()
        val move  = getLegalMoves(state).first { it.r == 6 && it.c == 4 && it.tr == 4 }
        val next  = applyMove(state, move)
        assertNotNull(next.enPassant)
        assertEquals(5, next.enPassant!!.r)
        assertEquals(4, next.enPassant!!.c)
    }

    @Test fun `applyMove flips turn`() {
        val state = newChessGame()
        val move  = getLegalMoves(state).first()
        assertEquals(CH_BLACK, applyMove(state, move).turn)
    }

    @Test fun `applyMove piece is moved on board`() {
        val state = newChessGame()
        val move  = getLegalMoves(state).first { it.r == 6 && it.c == 4 && it.tr == 4 }
        val next  = applyMove(state, move)
        assertEquals(0,        next.board[6][4])
        assertEquals(CH_PAWN,  next.board[4][4])
    }

    // ── isInCheck / isSquareAttacked ──────────────────────────────────────────

    @Test fun `white king is not in check at start`() {
        assertFalse(isInCheck(newChessGame().board, CH_WHITE))
    }

    @Test fun `black king is not in check at start`() {
        assertFalse(isInCheck(newChessGame().board, CH_BLACK))
    }

    @Test fun `isSquareAttacked detects rook attack`() {
        val board = Array(8) { IntArray(8) }
        board[0][0] = CH_ROOK  // white rook at a8
        assertTrue(isSquareAttacked(board, 0, 4, CH_WHITE))  // e8 attacked along rank
    }

    @Test fun `isSquareAttacked detects knight attack`() {
        val board = Array(8) { IntArray(8) }
        board[4][4] = CH_KNIGHT
        assertTrue(isSquareAttacked(board, 2, 3, CH_WHITE))
    }

    @Test fun `isSquareAttacked detects pawn attack`() {
        val board = Array(8) { IntArray(8) }
        board[5][4] = CH_PAWN  // white pawn at e3; attacks d2 and f2
        assertTrue(isSquareAttacked(board, 4, 3, CH_WHITE))
        assertTrue(isSquareAttacked(board, 4, 5, CH_WHITE))
    }

    // ── findKing ──────────────────────────────────────────────────────────────

    @Test fun `findKing locates white king at e1`() {
        val k = findKing(newChessGame().board, CH_WHITE)
        assertNotNull(k)
        assertEquals(7, k.r); assertEquals(4, k.c)
    }

    @Test fun `findKing locates black king at e8`() {
        val k = findKing(newChessGame().board, CH_BLACK)
        assertNotNull(k)
        assertEquals(0, k.r); assertEquals(4, k.c)
    }

    // ── updateStatus ──────────────────────────────────────────────────────────

    @Test fun `updateStatus returns legal moves`() {
        val updated = updateStatus(newChessGame())
        assertEquals(20, updated.legalMoves.size)
        assertEquals(ChessStatus.ACTIVE, updated.status)
    }

    @Test fun `updateStatus detects checkmate in scholar mate`() {
        // Set up a board position that is checkmate for black (Scholar's mate)
        var state = newChessGame()
        // e4
        state = applyMove(state, getLegalMoves(state).first { it.r==6 && it.c==4 && it.tr==4 })
        // e5
        state = applyMove(state, getLegalMoves(state).first { it.r==1 && it.c==4 && it.tr==3 })
        // Bc4
        state = applyMove(state, getLegalMoves(state).first { it.r==7 && it.c==5 && it.tr==4 && it.tc==2 })
        // Nc6
        state = applyMove(state, getLegalMoves(state).first { it.r==0 && it.c==1 && it.tr==2 && it.tc==2 })
        // Qh5
        state = applyMove(state, getLegalMoves(state).first { it.r==7 && it.c==3 && it.tr==3 && it.tc==7 })
        // Nf6?? (blunder)
        state = applyMove(state, getLegalMoves(state).first { it.r==0 && it.c==6 && it.tr==2 && it.tc==5 })
        // Qxf7#
        state = applyMove(state, getLegalMoves(state).first { it.r==3 && it.c==7 && it.tr==1 && it.tc==5 })
        val final = updateStatus(state)
        assertEquals(ChessStatus.CHECKMATE, final.status)
    }

    // ── evaluateBoard ─────────────────────────────────────────────────────────

    @Test fun `evaluateBoard returns 0 for starting position`() {
        assertEquals(0, evaluateBoard(newChessGame().board))
    }

    // ── getChessAIMove ────────────────────────────────────────────────────────

    @Test fun `getChessAIMove returns a legal move`() {
        val state  = updateStatus(newChessGame())
        val aiMove = getChessAIMove(state, difficulty = 2)
        assertNotNull(aiMove)
        assertTrue(state.legalMoves.contains(aiMove))
    }

    @Test fun `getChessAIMove returns null when no moves`() {
        val state = newChessGame().copy(
            board = Array(8) { IntArray(8) },
            legalMoves = emptyList(),
        )
        assertNull(getChessAIMove(state, 1))
    }

    @Test fun `castling rights revoked when king moves`() {
        val state = newChessGame()
        // Manually clear squares so king can move
        val board = state.copyBoard()
        board[7][5] = 0; board[7][6] = 0  // clear f1 and g1
        val s2 = state.copy(board = board)
        val kMove = getLegalMoves(s2).firstOrNull { it.r == 7 && it.c == 4 }
        assertNotNull(kMove)
        val next = applyMove(s2, kMove!!)
        assertFalse(next.castling.wK)
        assertFalse(next.castling.wQ)
    }
}
